### ISSUE TITLE
Can't resolve '../build/Debug/oracledb'

### ISSUE THREAD

**Issue by Adam95** [oracle/node-oracledb#534]
Hello,

today, all of a sudden (mid-work), I started getting the following error:
https://gist.github.com/Adam95/e96e4f4ac7f5ed781b10f98284a1fd2d
I presume it's a runtime error.

I'm using Ubuntu 16.04 LTS, 64bit.

I have the instant client installed via the rpm packages with:
```
sudo alien -i oracle-instantclient12.1-basic-12.1.0.2.0-1.x86_64.rpm
sudo alien -i oracle-instantclient12.1-devel-12.1.0.2.0-1.x86_64.rpm
```

This finishes without errors. Then, in my project folder, I run:
```
npm i oracledb
```
That finished as well. It is apparent from the output that it compiles to .o files and finally links them
to ../node_modules/oracledb/build/Release/oracledb.node

However when i run:
```
npm run watch:dev
```
which essentially runs webpack, I get the error in the gist above. Therefore, when I run 
`node src/server.js` and try to load the website in browser, I get 
`Error: NJS-045: cannot load the oracledb add-on binary`

As for environment variables:
https://gist.github.com/Adam95/d678815abbe5555fa39ace403acda4e2
(printed with `env`)

My node version is 4.2.6 which is not the latest but as I said, it worked fine before with this version.
If node-gyp (or what builds the library) uses my gcc, the it used gcc version 5.4.0

Only thing I can think of, which might caused this problem, was me installing sqldeveloper but I seriously doubt it because I installed it only by copying one folder to /opt/sqldeveloper and putting one binary to /usr/local/bin.

From my point of view (after I read many discussions here on github), it appears I must have some path/environment variables incorrectly set but I spent whole work-day by reinstalling and tweaking this that I am now seriously out of ideas.

Many thanks in advance,

Adam

**Comment by cjbj** [comment]
Do you need LD_LIBRARY_PATH set to the location of the Instant Client libraries?

**Comment by dmcghan** [comment]
@Adam95 Would you mind doing the following? Open ../node_modules/oracledb/lib/oracledb.js and add the following to line 33 (between the variable declarations and the try/catch blocks: 
```js
console.log(__dirname);
```

Then test your code again.

Does the output match where the try/catch blocks that follow look for oracledb?

Also, could you try a similar test that doesn't use webpack to see if it behaves differently?

**Comment by Adam95** [comment]
Thanks for the suggestion, I will try that in a few hours when I get home. The problem is surely within the web application or the configuration, the instant client works fine when I use sqlplus.

**Comment by Adam95** [comment]
@dmcghan So the `console.log(__dirname);` prints:
`/home/adam/......./node_modules/oracledb/lib`
Therefore the following search path in the try/catch block ('../build/Release/oracledb') is valid.

By different test than webpack you mean some different packing/compiling tool? I'm still a bit new to all these tools, so I used a starter project for angular2 which uses webpack. And it was working fine for the past week when I was using it.


@cjbj At this point I'm not really sure which environment variables I need. From the discussions here on GitHub I concluded that OCI_LIB_DIR and OCI_INC_DIR I no longer need after installing oracledb with npm. So do I need to point the LD_LIBRARY_PATH to instant client's lib directory?

I thought I could run `ldd` on oracledb/build/Release/oracledb.node and I got this output:
https://gist.github.com/Adam95/1b88c4e13b1b0dbf4b64cfd7e4c524a8

From the output it seems to me that the system already knows where the oracle libraries are, that's why I don't know if the LD_LIBRARY_PATH is needed.

**Comment by cjbj** [comment]
@Adam95 generally you will need `LD_LIBRARY_PATH` set, unless `FORCE_RPATH` is set (explicitly or implicity) during install.  Refer to the extensive installation guide https://github.com/oracle/node-oracledb/blob/master/INSTALL.md

**Comment by cjbj** [comment]
From checking your gist, the shell that you ran ldd in is fine.  This may or may not also be true for the shell that actually invokes node.  I suspect it is true, so the issue is elsewhere.

Do example programs run e.g. 'node examples/select1.js' ?

**Comment by Adam95** [comment]
This is very odd. If I create a new folder somewhere in my filesystem where I execute `npm install oracledb` and then run `node select1.js` it works fine.
I didn't set any environment variables like OCI_LIB_DIR etc. and oracledb still builds.

If I run `node src/server.js` which runs the backend part of my application, it runs fine and prints the __dirname into console with the command suggested by @dmcghan. So the server essentially runs but when I try to load the website in browser it shows: `Error: NJS-045: cannot load the oracledb add-on binary`

However when I run webpack which compiles my typescript angular2 frontend it shows the original error. So I should probably look for some clues here.

**Comment by Adam95** [comment]
OK, so I solved the issue. After countless hours of struggle I put the oracledb library as 'external' to webpack by putting following code in webpack.dev.js (development mode config)

```
var oracledblib = require('oracledb');
...
...
module.exports = {
  ...
  ...
  externals = [
    "oracledb": oracledblib
  ]
}
```

I still don't know why it broke down in the first place though.
Thank you again for support and quick replies.

Adam

**Comment by Steague** [comment]
@Adam95 && @cjbj I followed your example and added `oracledb` to the webpack externals. This got rid of the error at runtime but has now made it so the `oracledb` being imported in the packed app is `undefined`.

I tried copying the `oracledb` directory from the node_modules directory to webpack's `dist` directory but the `oracledb` module is still undefined. Any way you guys know how I can get the `oracledb` external to get picked up by the packed app?
